File: ts/src/session/runtime-session-read-model.ts
import type { RuntimeSessionEventLog } from "./runtime-events.js";
import { runtimeSessionIdForRun } from "./runtime-session-ids.js";

export interface RuntimeSessionSummary {
  session_id: string;
  parent_session_id: string;
  task_id: string;
  worker_id: string;
  goal: string;
  event_count: number;
  created_at: string;
  updated_at: string;
}

export interface RuntimeSessionReadStore {
  list(opts?: { limit?: number }): RuntimeSessionEventLog[];
  load(sessionId: string): RuntimeSessionEventLog | null;
}

export function summarizeRuntimeSession(log: RuntimeSessionEventLog): RuntimeSessionSummary {
  return {
    session_id: log.sessionId,
    parent_session_id: log.parentSessionId,
    task_id: log.taskId,
    worker_id: log.workerId,
    goal: readMetadataString(log.metadata, "goal"),
    event_count: log.events.length,
    created_at: log.createdAt,
    updated_at: log.updatedAt || log.createdAt,
  };
}

export function readRuntimeSessionSummaries(
  store: RuntimeSessionReadStore,
  opts: { limit?: number } = {},
): RuntimeSessionSummary[] {
  return store.list({ limit: opts.limit }).map(summarizeRuntimeSession);
}

export function readRuntimeSessionById(
  store: RuntimeSessionReadStore,
  sessionId: string,
): RuntimeSessionEventLog | null {
  return store.load(sessionId);
}

export function readRuntimeSessionByRunId(
  store: RuntimeSessionReadStore,
  runId: string,
): RuntimeSessionEventLog | null {
  return store.load(runtimeSessionIdForRun(runId));
}

function readMetadataString(
  metadata: Record<string, unknown>,
  key: string,
): string {
  const value = metadata[key];
  return typeof value === "string" ? value : "";
}

File: autocontext/src/autocontext/session/runtime_session_read_model.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Protocol

from autocontext.session.runtime_events import RuntimeSessionEventLog
from autocontext.session.runtime_session_ids import runtime_session_id_for_run

RuntimeSessionSummary = dict[str, str | int]


class RuntimeSessionReadStore(Protocol):
    def list(self, *, limit: int = 50) -> list[RuntimeSessionEventLog]: ...

    def load(self, session_id: str) -> RuntimeSessionEventLog | None: ...


def summarize_runtime_session(log: RuntimeSessionEventLog) -> RuntimeSessionSummary:
    return {
        "session_id": log.session_id,
        "parent_session_id": log.parent_session_id,
        "task_id": log.task_id,
        "worker_id": log.worker_id,
        "goal": _metadata_str(log.metadata, "goal"),
        "event_count": len(log.events),
        "created_at": log.created_at,
        "updated_at": log.updated_at or log.created_at,
    }


def read_runtime_session_summaries(
    store: RuntimeSessionReadStore | Mapping[str, RuntimeSessionEventLog],
    *,
    limit: int = 50,
) -> list[RuntimeSessionSummary]:
    return [summarize_runtime_session(log) for log in _list_logs(store, limit=limit)]


def read_runtime_session_by_id(
    store: RuntimeSessionReadStore | Mapping[str, RuntimeSessionEventLog],
    session_id: str,
) -> RuntimeSessionEventLog | None:
    if isinstance(store, Mapping):
        return store.get(session_id)
    return store.load(session_id)


def read_runtime_session_by_run_id(
    store: RuntimeSessionReadStore | Mapping[str, RuntimeSessionEventLog],
    run_id: str,
) -> RuntimeSessionEventLog | None:
    return read_runtime_session_by_id(store, runtime_session_id_for_run(run_id))


def _list_logs(
    store: RuntimeSessionReadStore | Mapping[str, RuntimeSessionEventLog],
    *,
    limit: int,
) -> list[RuntimeSessionEventLog]:
    if isinstance(store, Mapping):
        return list(store.values())[:limit]
    return store.list(limit=limit)


def _metadata_str(metadata: dict[str, Any], key: str) -> str:
    value = metadata.get(key)
    return value if isinstance(value, str) else ""
